rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala

```diff
@@ -2,185 +2,186 @@ package coop.rchain.rspace.test
+  private[rspace] override def addJoin(txn: T, c: C, cs: Seq[C]): Unit = {
+    val existing: Seq[Seq[C]] = _dbJoins.remove(c).getOrElse(Seq.empty)
+    if (!existing.exists(_.equals(cs)))
+      _dbJoins.put(c, cs +: existing)
+    else
+      _dbJoins.put(c, existing)
+  }
 
-  private[rspace] def removeDatum(txn: T, channels: Seq[C], index: Int): Unit = {
-    val key = hashChannels(channels)
-    for (as <- _data.get(key)) {
-      _data.update(key, dropIndex(as, index))
+  private[rspace] override def removeDatum(txn: T, channels: Seq[C], index: Int): Unit =
+    withGNAT(txn, hashChannels(channels)) { (_, _, gnatOpt) =>
+      gnatOpt.map(gnat => gnat.copy(data = dropIndex(gnat.data, index)))
     }
-    collectGarbage(key)
-  }
 
-  private[rspace] def removeWaitingContinuation(txn: T, channels: Seq[C], index: Int): Unit = {
+  private[rspace] override def removeWaitingContinuation(txn: T,
+                                                         channels: Seq[C],
+                                                         index: Int): Unit = {
     val key = hashChannels(channels)
-    for (psks <- _waitingContinuations.get(key)) {
-      _waitingContinuations.update(key, dropIndex(psks, index))
+    withGNAT(txn, key) { (_, _, gnatOpt) =>
+      gnatOpt.map(gnat => gnat.copy(wks = dropIndex(gnat.wks, index)))
     }
-    collectGarbage(key)
   }
 
-  private[rspace] def removeAll(txn: Unit, channels: Seq[C]): Unit = {
-    val key = hashChannels(channels)
-    _data.put(key, Seq.empty)
-    _waitingContinuations.put(key, Seq.empty)
+  private[rspace] override def removeAll(txn: Unit, channels: Seq[C]): Unit = {
+    withGNAT(txn, hashChannels(channels)) { (_, key, gnatOpt) =>
+      gnatOpt.map(_.copy(wks = Seq.empty, channels = Seq.empty))
+    }
     for (c <- channels) removeJoin(txn, c, channels)
   }
 
-  private[rspace] def addJoin(txn: T, c: C, cs: Seq[C]): Unit = {
-    val existing: Seq[Seq[C]] = _joinMap.remove(c).getOrElse(Seq.empty)
-    if (!existing.exists(_.equals(cs)))
-      _joinMap.put(c, cs +: existing)
-    else
-      _joinMap.put(c, existing)
-  }
-
-  private[rspace] def getJoin(txn: T, c: C): Seq[Seq[C]] =
-    _joinMap.getOrElse(c, Seq.empty[Seq[C]])
-
-  private[rspace] def removeJoin(txn: T, c: C, cs: Seq[C]): Unit = {
-    val joinKey = hashChannels(Seq(c))
-    val csKey   = hashChannels(cs)
-    if (_waitingContinuations.get(csKey).forall(_.isEmpty)) {
-      val existing: Seq[Seq[C]] = _joinMap.remove(c).getOrElse(Seq.empty)
-      val filtered              = existing.filter(!_.equals(cs))
-      if (filtered.nonEmpty)
-        _joinMap.put(c, filtered)
+  private[rspace] override def removeJoin(txn: T, c: C, cs: Seq[C]): Unit = {
+    withGNAT(txn, hashChannels(cs)) { (_, csKey, gnatOpt) =>
+      if (gnatOpt.isEmpty || gnatOpt.get.wks.isEmpty) {
+        val existing: Seq[Seq[C]] = _dbJoins.remove(c).getOrElse(Seq.empty)
+        val filtered              = existing.filter(!_.equals(cs))
```

Comment:
Looks like we have a divergence in implementations - in `LMDBStore` we only remove the first match.  Changing to that here keeps the tests passing.  The MultiMap would have prevented duplicate entries.  I'm inclined to say that `filter` is actually marginally more correct, although I'm trying to come up with a real-world example where duplicate `Seq[C]`'s would be in there.  We might consider trying to maintain the invariant of only unique `Seq`s and see what happens...

Reply:
will align with `LMDBStore`